src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs

```diff
@@ -0,0 +1,1677 @@
+                _isNotPrimitive = true;
+            }
+            else
+            {
+                // Create local copy to avoid bounds checks.
+                ReadOnlySpan<byte> localBuffer = _buffer;
+
+                if (JsonReaderHelper.IsDigit(first) || first == '-')
+                {
+                    if (!TryGetNumber(localBuffer.Slice(_consumed), out int numberOfBytes))
+                    {
+                        return false;
+                    }
+                    _tokenType = JsonTokenType.Number;
+                    _consumed += numberOfBytes;
+                    _bytePositionInLine += numberOfBytes;
+                    goto Done;
+                }
+                else if (ConsumeValue(first))
+                {
+                    goto Done;
+                }
+
+                return false;
+
+            Done:
+                // Cannot use HasMoreData since the JSON payload contains a single, non-primitive value
+                // and hence must be handled differently.
+                if (_consumed >= (uint)localBuffer.Length)
+                {
+                    return true;
+                }
+
+                if (localBuffer[_consumed] <= JsonConstants.Space)
+                {
+                    SkipWhiteSpace();
+                    if (_consumed >= (uint)localBuffer.Length)
+                    {
+                        return true;
+                    }
+                }
+
+                if (_readerOptions.CommentHandling != JsonCommentHandling.Disallow)
+                {
+                    if (_readerOptions.CommentHandling == JsonCommentHandling.Allow)
+                    {
+                        // This is necessary to avoid throwing when the user has 1 or more comments as the first token
+                        // OR if there is a comment after a single, non-primitive value.
+                        // In this mode, ConsumeValue consumes the comment and we need to return it as a token.
+                        // along with future comments in subsequeunt reads.
+                        if (_tokenType == JsonTokenType.Comment || localBuffer[_consumed] == JsonConstants.Slash)
+                        {
+                            return true;
+                        }
+                    }
+                    else
+                    {
+                        Debug.Assert(_readerOptions.CommentHandling == JsonCommentHandling.Skip);
+                        if (_tokenType == JsonTokenType.StartObject || _tokenType == JsonTokenType.StartArray)
+                        {
+                            _isNotPrimitive = true;
+                        }
+                        return true;
+                    }
+                }
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndAfterSingleJson, localBuffer[_consumed]);
+            }
+            return true;
+        }
+
+        private void SkipWhiteSpaceMultiSegment()
```

Comment:
This looks identical to the Span one. Am I missing something?

Come to think of it, a lot of these methods don't seem to talk about the sequence object, or call GetNextSpan, so they should be (seemingly) identical to the existing ones.

Reply:
Fixed. Removing WIP